unified-codes/tools/scripts/dgraph/migration/v2/import.js

```diff
@@ -0,0 +1,118 @@
+const excel = require('exceljs');
+const dgraph = require('dgraph-js');
+
+const inputFilename = process.argv[2] || './UC Spreadsheet 2.5.xlsm';
+
+const clientStub = new dgraph.DgraphClientStub(`localhost:9080`);
+const dgraphClient = new dgraph.DgraphClient(clientStub);
+
+const insertRootNodes = async () => {
+  const query = `query {
+    DrugCategory as var(func: eq(dgraph.type, Category)) @filter(eq(name, "Drug"))
+    ConsumableCategory as var(func: eq(dgraph.type, Category)) @filter(eq(name, "Consumable"))
+    OtherCategory as var(func: eq(dgraph.type, Category)) @filter(eq(name, "Other"))
+  }`;
+
+  const mutation = new dgraph.Mutation();
+  mutation.setSetNquads(`
+    uid(DrugCategory) <name> "Drug" .
+    uid(ConsumableCategory) <name> "Consumable" .
+    uid(OtherCategory) <name> "Other" .
+    uid(DrugCategory) <dgraph.type> "Category" .
+    uid(ConsumableCategory) <dgraph.type> "Category" .
+    uid(OtherCategory) <dgraph.type> "Category" .
+    uid(DrugCategory) <code> "933f3f00" .
+    uid(ConsumableCategory) <code> "77fcbb00" .
+  `);
+  // TODO: Source category codes from config or spreadsheet rather than hardcoding
+
+  const req = new dgraph.Request();
+  req.setQuery(query);
+  req.setMutationsList([mutation]);
+  req.setCommitNow(true);
+
+  const txn = dgraphClient.newTxn();
+  try {
+    await txn.doRequest(req);
+  } finally {
+    txn.discard();
+  }
+};
+
+const processRow = async (row) => {
+  const product = row.getCell(1).value;
+  const synonyms = row.getCell(2).value;
+  const categoryCode = row.getCell(11).value;
+  const productCode = row.getCell(12).value;
+
+  // Process non combination drugs first
+  if (!(product.indexOf('/') > -1)) {
+    const synonymList = synonyms ? synonyms.split(',') : [];
+    await insertProduct(product, productCode, categoryCode, synonymList);
+  }
+};
+
+const insertProduct = async (productName, productCode, categoryCode, synonymsArray) => {
+  const query = `query {
+    Category as var(func: eq(dgraph.type, Category)) @filter(eq(code, ${categoryCode}))
+    Product as var(func: eq(code, ${productCode})) 
+  }`;
+
+  let synonymMutation = '';
+  for (let i = 0; i < synonymsArray.length; i++) {
```

Comment:
perfect as it is. just for fun, also possible using `reduce()`

```
const synonymMutation = synonymsArray.reduce((mutation, synonym, i) => `${mutation}uid(Product) <name@alt${i}> "${synonym.trim()}" .\n`, '');

```

Reply:
Ooh fancy 👍  thanks for that